File: 0953-verifying-an-alien-dictionary/0953-verifying-an-alien-dictionary.py
class Solution:
    def isAlienSorted(self, words: List[str], order: str) -> bool:
        hashWords, charMap = [], {}

        for i in range(0, len(order)):
            charMap[order[i]] = i+1

        for w in range(0, len(words)):
            word, l, temp = words[w], 0, []

            while l < len(word):
                hashedLetter = charMap[word[l]]
                temp.append(hashedLetter)
                l += 1
            hashWords.append(temp)

        i, j = 1, 0
        
        while i < len(hashWords):
            wordA, wordB = hashWords[i-1], hashWords[i]
            equal = True

            while j < len(wordA) and j < len(wordB):
                if wordA[j] < wordB[j]:
                    equal = False
                    break
                elif wordA[j] > wordB[j]:
                    return False
                j += 1
            
            if equal and len(wordA) > len(wordB):
                return False
            i += 1

        return True

File: 0953-verifying-an-alien-dictionary/0953-verifying-an-alien-dictionary.cpp
static const bool Booster = [](){
    ios_base::sync_with_stdio(false);
    cout.tie(nullptr);
    cin.tie(nullptr);
    return true;
}();
class Solution {
public:
    bool isAlienSorted(vector<string>& words, const string order) {
        map<char, int> orderMap;
        int idx = 1;
        for (const char c : order) {
            orderMap[c] = idx;
            ++idx;
        }
        for (int i = 0; i < words.size() - 1; ++i) {
            string wordA = words[i];
            string wordB = words[i+1];
            int j = 0;
            bool substr = true;
            while (j < wordA.size() && j < wordB.size()) {
                if (orderMap[wordA[j]] < orderMap[wordB[j]]) {
                    substr = false;
                    break;
                }
                if (orderMap[wordA[j]] > orderMap[wordB[j]]) return false;
                if (substr && wordA[j] != wordB[j]) substr = false;
                ++j;
            }
            if (substr && wordA.size() > wordB.size()) return false;
        }
        return true;
    }
};;